pkg/index/inverted/index_filter.go

```diff
@@ -15,38 +15,129 @@
+func BuildLocalQuery(criteria *modelv1.Criteria, schema logical.Schema, entityDict map[string]int,
+	entity []*modelv1.TagValue,
+) (*BlugeQuery, [][]*modelv1.TagValue, bool, error) {
+	if criteria == nil {
+		return nil, [][]*modelv1.TagValue{entity}, false, nil
+	}
+	switch criteria.GetExp().(type) {
+	case *modelv1.Criteria_Condition:
+		cond := criteria.GetCondition()
+		expr, parsedEntity, err := parseExprOrEntity(entityDict, entity, cond)
+		if err != nil {
+			return nil, nil, false, err
+		}
+		if parsedEntity != nil {
+			return nil, parsedEntity, false, nil
+		}
+		if ok, indexRule := schema.IndexDefined(cond.Name); ok {
+			return parseConditionToQuery(cond, indexRule, expr, entity)
+		}
+		return nil, nil, false, errors.Wrapf(logical.ErrUnsupportedConditionOp, "mandatory index rule conf:%s", cond)
+	case *modelv1.Criteria_Le:
+		le := criteria.GetLe()
+		if le.GetLeft() == nil && le.GetRight() == nil {
+			return nil, nil, false, errors.WithMessagef(errInvalidLogicalExpression, "both sides(left and right) of [%v] are empty", criteria)
+		}
+		if le.GetLeft() == nil {
+			return BuildLocalQuery(le.Right, schema, entityDict, entity)
+		}
+		if le.GetRight() == nil {
+			return BuildLocalQuery(le.Left, schema, entityDict, entity)
+		}
+		left, leftEntities, leftIsMatchAllQuery, err := BuildLocalQuery(le.Left, schema, entityDict, entity)
+		if err != nil {
+			return nil, nil, false, err
+		}
+		right, rightEntities, rightIsMatchAllQuery, err := BuildLocalQuery(le.Right, schema, entityDict, entity)
+		if err != nil {
+			return nil, nil, false, err
+		}
+		entities := parseEntities(le.Op, entity, leftEntities, rightEntities)
+		if entities == nil {
+			return nil, nil, false, nil
+		}
+		if left == nil && right == nil {
+			return nil, entities, false, nil
+		}
+		if leftIsMatchAllQuery && rightIsMatchAllQuery {
+			return &BlugeQuery{bluge.NewMatchAllQuery()}, entities, true, nil
+		}
+		switch le.Op {
+		case modelv1.LogicalExpression_LOGICAL_OP_AND:
+			query := bluge.NewBooleanQuery()
+			if left != nil {
+				query.AddMust(left.Query)
+			}
+			if right != nil {
+				query.AddMust(right.Query)
+			}
+			return &BlugeQuery{query}, entities, false, nil
+		case modelv1.LogicalExpression_LOGICAL_OP_OR:
+			if leftIsMatchAllQuery || rightIsMatchAllQuery {
+				return &BlugeQuery{bluge.NewMatchAllQuery()}, entities, true, nil
+			}
+			query := bluge.NewBooleanQuery()
+			query.SetMinShould(1)
+			if left != nil {
+				query.AddShould(left.Query)
+			}
+			if right != nil {
+				query.AddShould(right.Query)
+			}
+			return &BlugeQuery{query}, entities, false, nil
+		}
+	}
+	return nil, nil, false, logical.ErrInvalidCriteriaType
+}
+
 // BuildLocalFilter returns a new index.Filter for local indices.
-func BuildLocalFilter(criteria *modelv1.Criteria, schema Schema, entityDict map[string]int,
-	entity []*modelv1.TagValue, mandatoryIndexRule bool,
+func BuildLocalFilter(criteria *modelv1.Criteria, schema logical.Schema,
```

Comment:
Why do you move the filter here? Its implementation doesn't relate to bulge.

It should be in the `pkg/query/logical/stream`